jobs/base_model_validator.py

```diff
@@ -0,0 +1,192 @@
+datastore_services = models.Registry.import_datastore_services()
+
+MAX_CLOCK_SKEW_SECS = datetime.timedelta(seconds=1)
+
+
+class BaseValidatorDoFn(beam.DoFn):
+    """"Base DoFn for model validations."""
+
+    def clone_model(self, model, **new_values):
+        """Clones the entity, adding or overriding constructor attributes.
+
+        The cloned entity will have exactly the same property values as the
+        original entity, except where overridden. By default, it will have no
+        parent entity or key name, unless supplied.
+
+        Args:
+            model: datastore_services.Model. Model to clone.
+            **new_values: dict(str: *). Keyword arguments to override when
+                invoking the cloned entity's constructor.
+
+        Returns:
+            *. A cloned, and possibly modified, copy of self. Subclasses of
+            BaseModel will return a clone with the same type.
+        """
+        # Reference implementation: https://stackoverflow.com/a/2712401/4859885.
+        cls = model.__class__
+        props = {k: v.__get__(model, cls) for k, v in cls._properties.items()} # pylint: disable=protected-access
+        props.update(new_values)
+        return cls(id=model.id, **props)
+
+
+class ValidateModelIdWithRegex(BaseValidatorDoFn):
+    """DoFn to validate model ids against a given regex string."""
+
+    def __init__(self, regex_string):
+        """Initializes the ValidateModelIdWithRegex DoFn.
+
+        Args:
+            regex_string: str. A regex pattern for valid ids.
+        """
+        super(ValidateModelIdWithRegex, self).__init__()
+        self.regex = re.compile(regex_string)
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            beam.pvalue.TaggedOutput. An element of the output PCollection for
+            the doFn which represents an error as a key value pair.
+        """
+        element = self.clone_model(model)
+
+        if not self.regex.match(element.id):
+            yield errors.ModelInvalidIdError(element)
+
+
+class ValidateDeleted(BaseValidatorDoFn):
+    """DoFn to check whether models marked for deletion are stale."""
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            beam.pvalue.TaggedOutput. An element of the output PCollection for
+            the doFn which represents an error as a key value pair.
+        """
+        element = self.clone_model(model)
+        date_now = datetime.datetime.utcnow()
+
+        date_before_which_models_should_be_deleted = (
+            date_now -
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED
+        )
```

Comment:
For the sake of consistency, let's avoid placing `)` on their own lines.

```suggestion
        date_before_which_models_should_be_deleted = (
            date_now -
            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED)
```